crates/verify/src/bytecode.rs

```diff
@@ -0,0 +1,653 @@
+    opts::EtherscanOpts,
+    utils::{self, read_constructor_args_file, LoadConfig},
+};
+use foundry_common::{
+    compile::{ProjectCompiler, SkipBuildFilter, SkipBuildFilters},
+    provider::alloy::ProviderBuilder,
+    types::ToEthers,
+};
+use foundry_compilers::{
+    artifacts::{BytecodeHash, BytecodeObject, CompactContractBytecode},
+    info::ContractInfo,
+    Artifact, EvmVersion,
+};
+use foundry_config::{figment, impl_figment_convert, Chain, Config};
+use foundry_evm::{
+    constants::DEFAULT_CREATE2_DEPLOYER, executors::TracingExecutor, utils::configure_tx_env,
+};
+use revm_primitives::{db::Database, EnvWithHandlerCfg, HandlerCfg, SpecId};
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use std::{fmt, path::PathBuf, str::FromStr};
+use yansi::Paint;
+
+impl_figment_convert!(VerifyBytecodeArgs);
+
+/// CLI arguments for `forge verify-bytecode`.
+#[derive(Clone, Debug, Parser)]
+pub struct VerifyBytecodeArgs {
+    /// The address of the contract to verify.
+    pub address: Address,
+
+    /// The contract identifier in the form `<path>:<contractname>`.
+    pub contract: ContractInfo,
+
+    /// The block at which the bytecode should be verified.
+    #[clap(long, value_name = "BLOCK")]
+    pub block: Option<BlockId>,
+
+    /// The constructor args to generate the creation code.
+    #[clap(
+        long,
+        conflicts_with = "constructor_args_path",
+        value_name = "ARGS",
+        visible_alias = "encoded-constructor-args"
+    )]
+    pub constructor_args: Option<String>,
+
+    /// The path to a file containing the constructor arguments.
+    #[clap(long, value_hint = ValueHint::FilePath, value_name = "PATH")]
+    pub constructor_args_path: Option<PathBuf>,
+
+    /// The rpc url to use for verification.
+    #[clap(short = 'r', long, value_name = "RPC_URL", env = "ETH_RPC_URL")]
+    pub rpc_url: Option<String>,
+
+    /// Verfication Type: `full` or `partial`. Ref: https://docs.sourcify.dev/docs/full-vs-partial-match/
+    #[clap(long, default_value = "full", value_name = "TYPE")]
+    pub verification_type: VerificationType,
+
+    #[clap(flatten)]
+    pub etherscan_opts: EtherscanOpts,
+
+    /// Skip building files whose names contain the given filter.
+    ///
+    /// `test` and `script` are aliases for `.t.sol` and `.s.sol`.
+    #[arg(long, num_args(1..))]
+    pub skip: Option<Vec<SkipBuildFilter>>,
+
+    /// The path to the project's root directory.
+    pub root: Option<PathBuf>,
+
+    /// Suppress logs and emit json results to stdout
+    #[clap(long, default_value = "false")]
+    pub json: bool,
+
+    /// The solc version of the cached artifact to use for verification in case of multiple
+    /// versions being present.
+    ///
+    /// If not provided, the most recent version is used.
+    ///
+    /// Can be specified in the format `x.y.z`, `solc:x.y.z` or `path/to/solc`.
+    #[clap(long, value_name = "CACHE_SOLC_VERSION")]
+    pub cache_version: Option<String>,
```

Comment:
do we need this? I believe etherscan has data on version used for verification

Reply:
We want to differentiate between what the user attempts to verify and what is onchain. 
Similar motivation https://github.com/foundry-rs/foundry/pull/7319#discussion_r1521981992

Reply:
On second thought, you're right. I got confused